sdk/src/transaction_context.rs

```diff
@@ -964,7 +964,10 @@ impl<'a> BorrowedAccount<'a> {
     /// in the given account. Does nothing if capacity is already sufficient.
     #[cfg(not(target_os = "solana"))]
     pub fn reserve(&mut self, additional: usize) -> Result<(), InstructionError> {
-        self.can_data_be_changed()?;
+        // Note that we don't need to call can_data_be_changed() here nor
+        // touch() the account. reserve() only changes the capacity of the
+        // memory that holds the account but it doesn't actually change content
+        // nor length of the account.
         self.make_data_mut();
         self.account.reserve(additional);
 
```

Comment:
(outside diff context)

this comment is stale in `fn make_data_mut()`:

```
// NOTE: The account memory region CoW code in Serializer::push_account_region() implements
// the same logic and must be kept in sync.
```

that function is now called `push_account_data_region`. moreover, seems the referred logic resides in a closure in `create_vm` now...

instead of comment which tend to bit-rot quite often, how about arranging code like this?

```rust
impl AccountSharedData {
    fn reserve(...) {
       ...
    }

    fn reserve_for_cow(&mut self) {
        if self.is_shared() {
            // if the account is still shared, it means this is the first time we're
            // about to write into it. Make the account mutable by copying it in a
            // buffer with MAX_PERMITTED_DATA_INCREASE capacity so that if the
            // transaction reallocs, we don't have to copy the whole account data a
            // second time to fullfill the realloc.

            self.reserve(MAX_PERMITTED_DATA_INCREASE);
        }
    }
}

impl BorrowedAccount {
    fn make_data_mut(&mut) {
        // the internally called `data_mut()` actually makes data mut.
        self.account.reserve_for_cow();
    }
}


  fn create_vm() ...
        || {
            ...
            account.reserve_for_cow();
            Ok(account.data_as_mut_slice().as_mut_ptr() as u64)
        },
```

Reply:
> instead of comment which tend to bit-rot quite often, how about arranging code like this?

Your suggestion breaks abstractions:

- The `reserve(MAX_PERMITTED_DATA_INCREASE)` optimization means: whenever a readable (shared) account needs to be written to *by a program*, ensure that it gets some spare capacity *for realloc*, so we don't have to allocate+memcpy again in case of realloc. This is done in `BorrowedAccount` which is the API we use when an account is accessed by programs within a transaction, where *realloc is possible and common*. The optimization is independent and even precedes CoW. CoW *benefits* from this optimization.

- `AccountSharedData` is the generic account API. At that level `MAX_PERMITTED_DATA_INCREASE` doesn't really make sense, and nor does CoW, they're both abstractions built _on top_ of `AccountSharedData`. Adding a new API to `AccountSharedData` that is specific to CoW and uses a constant that is only meaningful in another independent layer doesn't seem like a good idea to me.

I understand the general point of DRY, but this is literally two lines that are shared:

```
if account.is_shared() {
    account.reserve(MAX_PERMITTED_DATA_INCREASE);
 }
```

I'd rather use them in two call sites than add another not so great API method to AccountSharedData, especially considering that we've long decided that we're going to remove AccountSharedData entirely in PRv2.

Finally one thing to note is that this is a performance optimization only, even if for some extremely unlikely reason the optimization gets broken, it won't make anything incorrect.